### PR TITLE
[SYCL] Add implementation of kernel_bundle. Part 3

### PR DESCRIPTION
**Pull request by romanovvlad**
The patch adds
1. Implementation of more kernel_bundle APIs.
2. Support of kernel_bundles in handler

The patch duplicates some code in program manager in order to support
compilation for multiple devices. The refactoring of program manager
will be done in a separate patch.

Error handling will be added in a separate patch.

### CODE REVIEW

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 1]
```diff
@@ -36,18 +36,48 @@ namespace sycl {
...
-class queue_impl;
-} // namespace detail
 
 namespace detail {
 
+enum class ExtendedMembersType: unsigned int {
+  HANDLER_KERNEL_BUNDLE = 0,
+};
+
+struct ExtendedMember {
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/cg.hpp:46]: More comments are needed.

File: sycl/include/CL/sycl/kernel_bundle.hpp [hunk 2]
```diff
@@ -158,22 +158,6 @@ class __SYCL_EXPORT kernel_bundle_plain {
   bool native_specialization_constant() const noexcept;
 
 protected:
-  // \returns true if the kernel_bundle has the specialization constant with
-  // specified ID
-  bool has_specialization_constant(unsigned int SpecID) const noexcept;
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/kernel_bundle.hpp:163]: Removing spec consts related function to avoid having ABI breakage if it turns out they are not suitable.

File: sycl/source/detail/device_image_impl.hpp [hunk 3]
```diff
@@ -136,11 +145,43 @@ class device_image_impl {
...
+    return MSpecConstDescs;
+  }
+
+  OSModuleHandle get_OS_module_handle() const noexcept {
+    return MOSModuleHandle;
+  }
+
 private:
   RTDeviceBinaryImage *MBinImage = nullptr;
+  OSModuleHandle MOSModuleHandle = 0;
```
Comment (romanovvlad) [inline sycl/source/detail/device_image_impl.hpp:180]: Need to take it from sycl::get_kernel_bundle invocation.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 4]
```diff
@@ -1239,6 +1260,303 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
...
+  }
+
+  RTDeviceBinaryImage *ImgPtr = InputImpl->get_bin_image_ref();
+  const RTDeviceBinaryImage &Img = *ImgPtr;
+
+  auto BuildF = [this, &Context, Img, &Devs, &CompileOpts, &LinkOpts,
+                 &ExecImpl] {
+    // Update only if compile options are not overwritten by environment
+    // variable
+    if (!CompileOptsEnv) {
```
Comment (romanovvlad) [inline sycl/source/detail/program_manager/program_manager.cpp:1416]: These code partially overlaps with what program_manager already does in other functions. But those functions do not work with multiple devices.
Changing whole program_manager in scope of this patch is too intrusive, so going to refactor it separately.

File: sycl/source/handler.cpp [hunk 5]
```diff
@@ -14,23 +14,103 @@
...
   // It is harmless (does nothing) for everything else.
   if (MIsFinalized)
     return MLastEvent;
   MIsFinalized = true;
 
+  if (getCGTypeVersion(MCGType) > detail::CG::CG_VERSION::V0) {
+    // If there were uses of set_specialization_constant build the kernel_bundle
+    std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr =
+        getOrInsertHandlerKernelBundle(/*Insert=*/false);
+    if (KernelBundleImpPtr) {
```
Comment (romanovvlad) [inline sycl/source/handler.cpp:90]: The kernel_bundle is used if there are calls to use_kernel_bundle or set_specialization_constants only.

File: sycl/test/on-device/basic_tests/kernel_bundle/kernel_bundle_api.cpp [hunk 6]
```diff
@@ -95,6 +99,50 @@ int main() {
...
+        KernelBundle5.get_kernel_ids();
+
+    assert(KernelBundle5KernelIDs.size() == 2);
+
+    sycl::kernel_bundle KernelBundle6 =
+        sycl::get_kernel_bundle<sycl::bundle_state::input>(Ctx, {Dev});
+
+    sycl::kernel_bundle<sycl::bundle_state::object> KernelBundle6Compiled =
+        sycl::compile(KernelBundle6, KernelBundle6.get_devices());
+
```
Comment (romanovvlad) [inline sycl/test/on-device/basic_tests/kernel_bundle/kernel_bundle_api.cpp:128]: Add checks for PI calls.

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 7]
```diff
@@ -36,18 +36,48 @@ namespace sycl {
...
+      std::static_pointer_cast<const std::vector<ExtendedMember>>(SPtr));
+}
+
 class stream_impl;
+class queue_impl;
+class kernel_bundle_impl;
+
+constexpr static unsigned int getVersionedCGType(unsigned int Type,
+                                                 unsigned char Version) {
+  return Type | ((unsigned int)Version << 24);
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg.hpp:63]: What is 24?

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 8]
```diff
@@ -36,18 +36,48 @@ namespace sycl {
...
+
+constexpr static unsigned char getCGTypeVersion(unsigned int Type) {
+  return Type >> 24;
+}
+
 /// Base class for all types of command groups.
 class CG {
 public:
+
+  enum CG_VERSION: unsigned char {
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg.hpp:74]:
```suggestion
  enum class CG_VERSION: unsigned char {
```

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 4 above)
Comment (alexbatashev) [reply sycl/source/detail/program_manager/program_manager.cpp:1416]: Does it make sense to add a TODO comment?

File: sycl/source/handler.cpp [hunk 9]
```diff
@@ -14,23 +14,103 @@
...
 #include <detail/queue_impl.hpp>
 #include <detail/scheduler/scheduler.hpp>
 
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 
+context handler::getContext() { return MQueue->get_context(); }
+
+std::shared_ptr<detail::kernel_bundle_impl>
+handler::getOrInsertHandlerKernelBundle(bool Insert) {
```
Comment (alexbatashev) [inline sycl/source/handler.cpp:28]: Probably not worth it, but wouldn't it be better to have different functions for get and insert?

File: sycl/test/on-device/basic_tests/kernel_bundle/kernel_bundle_api.cpp [hunk 10]
```diff
@@ -5,11 +5,13 @@
 // UNSUPPORTED: cuda
 
 #include <CL/sycl.hpp>
+#include <detail/device_image_impl.hpp>
```
Comment (alexbatashev) [inline sycl/test/on-device/basic_tests/kernel_bundle/kernel_bundle_api.cpp:8]: I'm not sure you can use this include in a lit test.

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 11]
```diff
@@ -36,18 +36,48 @@ namespace sycl {
...
+convertToExtendedMembers(const std::shared_ptr<const void> &SPtr) {
+  return std::const_pointer_cast<std::vector<ExtendedMember>>(
+      std::static_pointer_cast<const std::vector<ExtendedMember>>(SPtr));
+}
+
 class stream_impl;
+class queue_impl;
+class kernel_bundle_impl;
+
+constexpr static unsigned int getVersionedCGType(unsigned int Type,
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg.hpp:61]: I may be wrong, but `constexpr` used in a function implies `inline`, thus you don't need static.
```suggestion
constexpr unsigned int getVersionedCGType(unsigned int Type,
```

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 7 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:63]: Added a comment.

File: sycl/source/handler.cpp (hunk 9 above)
Comment (romanovvlad) [reply sycl/source/handler.cpp:28]: In this case insert function would duplicate logic of get function because the whole operation should be done under mutex.

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 12]
```diff
@@ -36,18 +36,66 @@ namespace sycl {
...
+// Used to represent a type of an extended member
+enum class ExtendedMembersType: unsigned int {
+  HANDLER_KERNEL_BUNDLE = 0,
+};
+
+// Holds a pointer to an object of an arbitrary type and an ID value which
+// should be used to understand what type pointer points to.
+// Used as to extend handler class without introducing new class members which
+// would change handler layout.
+struct ExtendedMemberT {
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg.hpp:51]: As I can guess, this is aimed at solving some handler class backward compatibility problem. Is there some more detailed description elsewhere of the problem and solution? If not, it would be highly useful for the review and further maintenance to have it here. I don't fully understand what e.g. `convertToExtendedMembers` does.

Few more questions which would be good to address in comments:
- I assume there is some correspondence between a CG version and the set of extended members
- I assume changing the set of extended members should lead to CG version change

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 13]
```diff
@@ -36,18 +36,66 @@ namespace sycl {
...
+// Returns the version encoded to the type
+constexpr unsigned char getCGTypeVersion(unsigned int Type) {
+  return Type >> ShiftBitsForVersion;
+}
+
 /// Base class for all types of command groups.
 class CG {
 public:
+
+  enum class CG_VERSION: unsigned char {
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg.hpp:92]: Nit: 128 versions should probably be enough, but this discords with other enums. If there is perf/size consideration behind unsigned char selection - please add comment.

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 14]
```diff
@@ -62,7 +110,9 @@ class CG {
     FILL_USM = 11,
     PREFETCH_USM = 12,
     CODEPLAY_INTEROP_TASK = 13,
-    CODEPLAY_HOST_TASK = 14
+    CODEPLAY_HOST_TASK = 14,
+    KERNEL_V2 =
+        getVersionedCGType(KERNEL, static_cast<unsigned int>(CG_VERSION::V1)),
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg.hpp:115]: Nit: matching two versions would be more intuitive
```suggestion
    KERNEL_V1 =
        getVersionedCGType(KERNEL, static_cast<unsigned int>(CG_VERSION::V1)),
```

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 15]
```diff
@@ -89,6 +139,14 @@ class CG {
 
   CGTYPE getType() { return MType; }
 
+  std::shared_ptr<std::vector<ExtendedMemberT>> getExtendedMembers() {
+    if (getCGTypeVersion(MType) == static_cast<unsigned int>(CG_VERSION::V0) ||
+        MSharedPtrStorage.empty())
+      return nullptr;
+
+    return convertToExtendedMembers(MSharedPtrStorage[0]);
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg.hpp:147]: Not clear why [0]. Can't CG contain other shared ptrs? Please clarify

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 16]
```diff
@@ -36,18 +36,66 @@ namespace sycl {
...
+// Holds a pointer to an object of an arbitrary type and an ID value which
+// should be used to understand what type pointer points to.
+// Used as to extend handler class without introducing new class members which
+// would change handler layout.
+struct ExtendedMemberT {
+  ExtendedMembersType MType;
+  std::shared_ptr<void> MData;
+};
+
+static std::shared_ptr<std::vector<ExtendedMemberT>>
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/detail/cg.hpp:56]: NIT: why not simple pointer? I wonder if the result of convertToExtendedMembers can escape and get referenced by user structures?

File: sycl/include/CL/sycl/handler.hpp [hunk 17]
```diff
@@ -313,8 +314,7 @@ class __SYCL_EXPORT handler {
   ///
   /// \param Queue is a SYCL queue.
   /// \param IsHost indicates if this handler is created for SYCL host device.
-  handler(shared_ptr_class<detail::queue_impl> Queue, bool IsHost)
-      : MQueue(std::move(Queue)), MIsHost(IsHost) {}
+  handler(const shared_ptr_class<detail::queue_impl> &Queue, bool IsHost);
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:317]: NIT: SYCL2020 deprecates C++ std lib class aliases, such as shared_ptr_class, also std::shared_ptr is used in other RT places. So I `const std::shared_ptr<detail::queue_impl>` would be nicer, since Queue type is being changed anyway.

File: sycl/include/CL/sycl/handler.hpp [hunk 18]
```diff
@@ -900,12 +900,38 @@ class __SYCL_EXPORT handler {
...
   handler(handler &&) = delete;
   handler &operator=(const handler &) = delete;
   handler &operator=(handler &&) = delete;
 
+#if __cplusplus > 201402L
+  template <auto &SpecName>
+  void set_specialization_constant(
+      typename std::remove_reference_t<decltype(SpecName)>::type Value) {
+
+    std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr =
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/handler.hpp:920]: nit for consistency
```suggestion
    std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImplPtr =
```

File: sycl/include/CL/sycl/kernel_bundle.hpp [hunk 19]
```diff
@@ -340,7 +321,7 @@ namespace detail {
 // public onces
 __SYCL_EXPORT detail::KernelBundleImplPtr
 get_kernel_bundle_impl(const context &Ctx, const std::vector<device> &Devs,
-                       bundle_state State);
+                       bundle_state State, OSModuleHandle M);
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/kernel_bundle.hpp:324]: as agreed, OSModuleHandles can now be removed - conflicting kernel names in different OS modules should be  an error (according to @gmlueck). Please also add a TODO to diagnose conflicting kernel names.

File: sycl/include/CL/sycl/kernel_bundle.hpp [hunk 20]
```diff
@@ -503,7 +497,8 @@ bool has_kernel_bundle(const context &Ctx, const std::vector<device> &Devs) {
 
 /// \returns true if all of the kernels identified by KernelIDs are compatible
 /// with the device Dev.
-bool is_compatible(const std::vector<kernel_id> &KernelIDs, const device &Dev);
+__SYCL_EXPORT bool is_compatible(const std::vector<kernel_id> &KernelIDs,
```
Comment (kbobrovs) [inline sycl/include/CL/sycl/kernel_bundle.hpp:500]: NIT: I see some changes like this that do not really relate to kernel bundles, should better be separate PRs

**Review by kbobrovs — COMMENTED** [review]
1st part of the review (up to kernel_bundle.hpp)

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 13 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:92]: Added comment.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 12 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:51]: Added a comment.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 15 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:147]: Tried to explain it in the comment on top of the file.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 16 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:56]: Hypothetically it can. It seems logical to not provide interfaces using which naively one can get very hard to debug problems .

File: sycl/include/CL/sycl/handler.hpp (hunk 17 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/handler.hpp:317]: If we need to do it, it should be done in a lot of places. Is it OK if we do in a separate patch?

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 4 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:1395]: Added

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 21]
```diff
@@ -36,18 +36,113 @@ namespace sycl {
...
+  return std::const_pointer_cast<std::vector<ExtendedMemberT>>(
+      std::static_pointer_cast<const std::vector<ExtendedMemberT>>(SPtr));
+}
+
 class stream_impl;
+class queue_impl;
+class kernel_bundle_impl;
+
+// The constant is used to shift left CG type value to access it's version
+constexpr unsigned int ShiftBitsForVersion = 24;
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg.hpp:118]: Nit: I think this needs `static`, since it's not a function or a static member variable, and `inline` variables are not even a thing prior to C++17
```suggestion
static constexpr unsigned int ShiftBitsForVersion = 24;
```

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 22]
```diff
@@ -36,18 +36,113 @@ namespace sycl {
...
+
+// Constructs versioned type
+constexpr unsigned int getVersionedCGType(unsigned int Type,
+                                          unsigned char Version) {
+  return Type | ((unsigned int)Version << ShiftBitsForVersion);
```
Comment (alexbatashev) [inline sycl/include/CL/sycl/detail/cg.hpp:123]: Nit
```suggestion
  return Type | (static_cast<unsigned int>(Version) << ShiftBitsForVersion);
```

File: sycl/source/detail/device_image_impl.hpp [hunk 23]
```diff
@@ -136,11 +117,46 @@ class device_image_impl {
 
   void set_state(bundle_state NewState) noexcept { MState = NewState; }
 
+  std::vector<device> &get_devices() { return MDevices; }
```
Comment (alexbatashev) [inline sycl/source/detail/device_image_impl.hpp:120]:
```suggestion
  std::vector<device> &get_devices() noexcept { return MDevices; }
```

File: sycl/source/detail/device_image_impl.hpp [hunk 24]
```diff
@@ -136,11 +117,46 @@ class device_image_impl {
...
+
+  const std::vector<device> &get_devices() const { return MDevices; }
```
Comment (alexbatashev) [inline sycl/source/detail/device_image_impl.hpp:122]:
```suggestion
  const std::vector<device> &get_devices() const noexcept { return MDevices; }
```

**Review by alexbatashev — COMMENTED** [review]
Just a few nits

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 21 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:118]: It seems there is no difference. constexpr variable at namepspace scope has internal linkage anyway.

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 25]
```diff
@@ -62,7 +163,9 @@ class CG {
     FILL_USM = 11,
     PREFETCH_USM = 12,
     CODEPLAY_INTEROP_TASK = 13,
-    CODEPLAY_HOST_TASK = 14
+    CODEPLAY_HOST_TASK = 14,
+    KERNEL_V1 =
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/cg.hpp:167]: Probably it makes more sense to not have distinct enum value and hide manipulations with version in setType, getType, setVersion, getVersion methods.

File: sycl/include/CL/sycl/handler.hpp [hunk 26]
```diff
@@ -898,12 +898,38 @@ class __SYCL_EXPORT handler {
...
 public:
   handler(const handler &) = delete;
   handler(handler &&) = delete;
   handler &operator=(const handler &) = delete;
   handler &operator=(handler &&) = delete;
 
+#if __cplusplus > 201402L
+  template <auto &SpecName>
+  void set_specialization_constant(
+      typename std::remove_reference_t<decltype(SpecName)>::type Value) {
```
Comment (gmlueck) [inline sycl/include/CL/sycl/handler.hpp:916]: Note that the SYCL 2020 spec had a type.  The name `type` should instead be `value_type`.  See https://github.com/KhronosGroup/SYCL-Docs/pull/116

```suggestion
      typename std::remove_reference_t<decltype(SpecName)>::value_type Value) {
```

File: sycl/include/CL/sycl/handler.hpp [hunk 27]
```diff
@@ -898,12 +898,38 @@ class __SYCL_EXPORT handler {
...
+
+    std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImplPtr =
+        getOrInsertHandlerKernelBundle(/*Insert=*/true);
+
+    detail::createSyclObjFromImpl<kernel_bundle<bundle_state::input>>(
+        KernelBundleImplPtr)
+        .set_specialization_constant<SpecName>(Value);
+  }
```
Comment (gmlueck) [inline sycl/include/CL/sycl/handler.hpp:924]: Did you mean to add an implementation for `get_specialization_constant()` also?

File: sycl/include/CL/sycl/kernel_bundle.hpp [hunk 28]
```diff
@@ -572,16 +555,25 @@ compile(const kernel_bundle<bundle_state::input> &InputBundle,
...
+          const std::vector<device> &Devs, const property_list &PropList);
 }
 
 /// \returns a new kernel_bundle which contains the device images from the
 /// ObjectBundles that are translated into one or more new device images of
 /// state bundle_state::executable The new bundle represents all of the kernels
 /// in ObjectBundles that are compatible with at least one of the devices in
 /// Devs.
-kernel_bundle<bundle_state::executable>
+inline kernel_bundle<bundle_state::executable>
```
Comment (gmlueck) [inline sycl/include/CL/sycl/kernel_bundle.hpp:569]: Just curious, why are `compile()`, `link()`, and `build()` declared `inline` while none of the other functions in this header use `inline`?

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 29]
```diff
@@ -327,12 +328,17 @@ RT::PiProgram ProgramManager::createPIProgram(const RTDeviceBinaryImage &Img,
...
+         "Creating a program from AOT binary for multiple device is not "
+         "supported");
+
   // Load the image
   const ContextImplPtr Ctx = getSyclObjImpl(Context);
   RT::PiProgram Res =
       Format == PI_DEVICE_BINARY_TYPE_SPIRV
           ? createSpirvProgram(Ctx, RawImg.BinaryStart, ImgSize)
-          : createBinaryProgram(Ctx, Device, RawImg.BinaryStart, ImgSize);
+          : createBinaryProgram(Ctx, Devices[0], RawImg.BinaryStart, ImgSize);
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:341]: It seems no devices beyond index 0 are ever accessed. Why changing function signature then? If this is for some future extension - please add a comment then.

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 30]
```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
         BinImages.push_back(ImageUPtr.get());
     }
   }
+  // TODO: Add a diagnostic on multiple device images with conflicting kernel
+  // names.
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:1202]:
```suggestion
  // names, and remove OSModuleHandle usage, as conflicting kernel names will be an error.
```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 31]
```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
...
-    // with a target state if an image is already in the target state or can be
-    // brought to target state by compiling/linking/building.
-    //
-    // Example: an image in "executable" state is not compatbile with "input"
-    // target state - there is no operation to convert the image it to "input"
-    // state.
-    // An image in "input" state is compatible with "executable" target state
-    // because it can be built to get into "executable" state.
-    if (ImgState > TargetState)
+    if (ImgState != TargetState)
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:1210]: this filtering should better be done when populating BinImages

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 32]
```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
...
       continue;
 
     for (const sycl::device &Dev : Devs)
       if (compatibleWithDevice(BinImage, Dev)) {
+
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:1215]:
```suggestion
      if (!compatibleWithDevice(BinImage, Dev)) continue;

```

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 33]
```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
...
-  case bundle_state::executable:
-    assert(false && "Not implemented yet");
-    break;
+  return SYCLDeviceImages;
+}
+
+device_image_plain
+ProgramManager::compile(const device_image_plain &DeviceImage,
+                        const std::vector<device> &Devs,
+                        const property_list &) {
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:1248]: I don't see any way to pass options to compilation or linkage below. Is this a TODO? please add if so

File: sycl/source/detail/program_manager/program_manager.cpp [hunk 34]
```diff
@@ -1192,51 +1198,326 @@ ProgramManager::getSYCLDeviceImages(const context &Ctx,
...
+      std::make_shared<detail::device_image_impl>(
+          /*BinImage=*/nullptr, Context, Devs, bundle_state::object,
+          std::move(KernelIDs), LinkedProg);
+
+  // TODO: Make multiple sets of device images organized by devices they are
+  // compiled for.
+  return {createSyclObjFromImpl<device_image_plain>(ExecutableImpl)};
+}
+
+device_image_plain ProgramManager::build(const device_image_plain &DeviceImage,
```
Comment (kbobrovs) [inline sycl/source/detail/program_manager/program_manager.cpp:1334]: this is a big piece of highly complex code, most of it seems to be copy-paste from existing program-based build interface. could you please comment what is changed compared to the existing impl for faster review (or better mark with comments right here)?

**Review by kbobrovs — COMMENTED** [review]
Reviewed up to program manager (including).

**Review by alexbatashev — COMMENTED** [review]
LGTM if the comments above are resolved

File: sycl/include/CL/sycl/kernel_bundle.hpp (hunk 28 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/kernel_bundle.hpp:569]: The non-template functions which have definition in the header itself are marked with `inline` to avoid multiple definitions error.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 29 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:341]: They are accessed when creating from SPIRV. I hope assert clarifies it a bit.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 33 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:1248]: It is not defined in the spec, but it looks like intention is to pass them using property_list.
Not sure which TODO is needed here.

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 35]
```diff
@@ -68,6 +140,27 @@ class kernel_bundle_impl {
     MDeviceImages.erase(It, MDeviceImages.end());
   }
 
+  // C'tor matches sycl::join API
+  kernel_bundle_impl(const std::vector<detail::KernelBundleImplPtr> &Bundles) {
+    MContext = Bundles[0]->MContext;
+    for (const detail::KernelBundleImplPtr &Bundle : Bundles) {
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/source/detail/kernel_bundle_impl.hpp:146]: Probably I didn't get the idea. But if I understand it correctly it falls to the same problem we discussed previously - that using std::set here would require repacking values back to vector since there are required APIs which return values in std::vector.

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 36]
```diff
@@ -88,19 +181,42 @@ class kernel_bundle_impl {
 
       Result.insert(Result.end(), KernelIDs.begin(), KernelIDs.end());
     }
-    std::sort(Result.begin(), Result.end(),
-              [](const kernel_id &LHS, const kernel_id &RHS) {
-                return detail::getSyclObjImpl(LHS) <
-                       detail::getSyclObjImpl(RHS);
-              });
-    std::unique(Result.begin(), Result.end());
+    std::sort(Result.begin(), Result.end(), LessByNameComp{});
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/source/detail/kernel_bundle_impl.hpp:184]: Each device_image has kernel_ids sorted.But here we need to get merged list of kernel ids from all the device images. There is an option to use inplace_merge to merge several sorted vectors which was used initially, but I've got a comment that using regular sort should be not slower and more clear for code readers.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 34 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:1334]: Added a comment which outlines main differences. 
Adding it here as well:
The function duplicates most of the code from existing getBuiltPIProgram.
The differences are:
 Different API - uses different objects to extract required info
 Supports caching of a program built for multiple devices

File: sycl/source/detail/kernel_bundle_impl.hpp [hunk 37]
```diff
@@ -26,6 +27,12 @@ __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 namespace detail {
 
+template <class T> struct LessByHash {
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/source/detail/kernel_bundle_impl.hpp:30]: Right, but it seems we have no util header in the source dir. Does it make sense to add it to the one in include dir?The cons here is that it would be compiled with the user code while not needed.

File: sycl/include/CL/sycl/detail/cg.hpp (hunk 15 above)
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:202]: Yes, the idea is to avoid adding a new field to cg and handler since adding them breaks ABI.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 29 above)
Comment (kbobrovs) [reply sycl/source/detail/program_manager/program_manager.cpp:346]: Hm. `Devices` parameter is used 3 times in this function, as far as I can see: 
```cpp
for (const device &Dev : Devices) { // debug print Dev }
assert((Format == PI_DEVICE_BINARY_TYPE_SPIRV || Devices.size() == 1)
createBinaryProgram(Ctx, Devices[0], RawImg.BinaryStart, ImgSize);
```
So I don't see how any other device is accessed. Am I missing somthing?

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 33 above)
Comment (kbobrovs) [reply sycl/source/detail/program_manager/program_manager.cpp:1370]: So, the `property_list` parameter should serve this purpose, now I get it. But it is not used and I can see `/*options=*/nullptr`. So I'd suggest to add something like:
```cpp
// TODO Add translation of the property_list parameter to compilation options and pass them to `piProgramCompile`
```

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 34 above)
Comment (kbobrovs) [reply sycl/source/detail/program_manager/program_manager.cpp:1460]: thanks!

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 33 above)
Comment (gmlueck) [reply sycl/source/detail/program_manager/program_manager.cpp:1370]: If we want to support any compilation options passed via the property list, we would need to update the Level Zero backend specification (or perhaps some other extension spec) to document what those options are.  Until we do that, there is nothing TODO in the code here.  :-)

File: sycl/source/detail/kernel_bundle_impl.hpp (hunk 36 above)
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/source/detail/kernel_bundle_impl.hpp:172]: Will do in a separate PR.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 33 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:1370]: @gmlueck Could you please clarify how compile and link options can be passed in SYCL2020 API?

Comment (gmlueck) [reply sycl/source/detail/program_manager/program_manager.cpp:1370]: The definitions of `compile()`, `link()`, and `build()` all take a `property_list` parameter, but the core spec doesn't define any properties.  Instead, this is an extension point for vendors and backends.  Let's suppose that we wanted to expose the "-ze-opt-large-register-file" build option from Level Zero.  We would specify an extension that defines a new property like:

```
namespace sycl::ext::oneapi::level_zero {

namespace property {
struct large_register_file {};
// namespace property

static constexpr property::large_register_file large_register_file;

} // namespace sycl::ext::oneapi::level_zero
```

Applications could then build with this option like this:

```
build(inputBundle, {ext::oneapi::level_zero::large_register_file});
```

We could decide whether to expose this as a DPC++ extension or as part of the Level Zero Backend spec.  Assuming that these compilation options are considered part of the Level Zero API, it probably makes sense to expose the SYCL integration as part of the Level Zero backend spec.

Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp:1359]: Added a comment.

File: sycl/source/detail/program_manager/program_manager.cpp (hunk 29 above)
Comment (romanovvlad) [reply sycl/source/detail/program_manager/program_manager.cpp]: Reverted this change.

**Review by kbobrovs — COMMENTED** [review]
LGTM

**Review by kbobrovs — APPROVED** [review]
LGTM

File: sycl/include/CL/sycl/detail/cg.hpp [hunk 38]
```diff
@@ -36,18 +36,119 @@ namespace sycl {
...
+//   /\
+//   ||
+// The byte specifies the version
+//
+// A user of this vector should not expect that a specific data is stored at a
+// specific position, but iterate over all looking for an ExtendedMemberT value
+// with the desired type.
+// This allows changing/extending the contents of this vector without changing
+// the version.
+//
```
[the remark this answers is not in the document]
Comment (romanovvlad) [reply sycl/include/CL/sycl/detail/cg.hpp:91]: Ok, will add in a separate PR.
